src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs

```diff
@@ -82,13 +86,43 @@ internal InfiniteScrollList(Lazy<JoinableTaskFactory> joinableTaskFactory)
 
             BindingOperations.EnableCollectionSynchronization(Items, _list.ItemsLock);
 
-            ItemsView = new CollectionViewSource() { Source = Items }.View;
-            DataContext = ItemsView;
+            ItemsView = new CollectionViewSource() { Source = Items };
+            ListCollectionView itemsView = ItemsView.View as ListCollectionView;
+            itemsView.IsLiveFiltering = true;
+            itemsView.IsLiveGrouping = true;
+            itemsView.LiveFilteringProperties.Add($"{nameof(PackageItemViewModel.IsPackageVulnerable)}");
+            itemsView.LiveGroupingProperties.Add($"{nameof(PackageItemViewModel.PackageLevel)}");
+            itemsView.Filter = item =>
+            {
+                if (item is not null && item.Equals(_loadingVulnerabilitiesStatusIndicator))
+                {
+                    if (FilterByVulnerabilities)
+                    {
+                        return true;
+                    }
+                    return false;
+                }
+
+                PackageItemViewModel vitem = item as PackageItemViewModel;
+                return IsItemInFilter(vitem);
+            };
```

Comment:
Rewrite this to delegate to two methods, one for filtering on the loading indicator, and one for the vulnerable packages. That way, each filter method does one thing and one thing well, rather than mixing the logic of checking for the vulnerability loading indicator and the vulnerable package into a single method.

Disclaimer: Please test the boolean logic. I did not test this, but this is the structure I recommend, and I believe the logic to be equivalent to what was in the PR.

```suggestion
            itemsView.Filter = item =>
            {
                return FilterLoadingIndicator(item)
                    && FilterVulnerablePackage(item);
            };
```

Then add two methods:

```
        private bool FilterLoadingIndicator(object item)
        {
            if (item is not null && item.Equals(_loadingVulnerabilitiesStatusIndicator))
            {
                return FilterByVulnerabilities;
            }

            return true;
        }

        private bool FilterVulnerablePackage(object item)
        {
            return !FilterByVulnerabilities
                || (FilterByVulnerabilities && item is PackageItemViewModel vm && vm.IsPackageVulnerable);
        }
```

Reply:
This wasn't working for all scenarios and it was hard to understand where the bug was

Reply:
@jebriede I did something a little bit different but use most of it, doing `FilterLoadingIndicator(item) && FilterVulnerablePackage(item)` was causing some weird behavior, I decided to rely on if's instead